### PR TITLE
feat: Add API `TraverseStateChanges` to extract state changes from iavl versions

### PR DESCRIPTION
**Pull request by yihuang**
the diff algorithm is based on: https://github.com/cosmos/iavl/pull/646

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -9,6 +9,7 @@
 - [#586](https://github.com/cosmos/iavl/pull/586) Remove the `RangeProof` and refactor the ics23_proof to use the internal methods.
 - [#640](https://github.com/cosmos/iavl/pull/640) commit `NodeDB` batch in `LoadVersionForOverwriting`.
 - [#636](https://github.com/cosmos/iavl/pull/636) Speed up rollback method: `LoadVersionForOverwriting`.
+- [#]() Add API `TraverseStateChanges` to extract state changes from iavl versions.
```
Comment (yihuang) [inline CHANGELOG.md:12]:
```suggestion
- [#654](https://github.com/cosmos/iavl/pull/654) Add API `TraverseStateChanges` to extract state changes from iavl versions.
```

File: nodedb.go [hunk 2]
```diff
@@ -1042,6 +1042,39 @@ func (ndb *nodeDB) traverseNodes(fn func(hash []byte, node *Node) error) error {
 	return nil
 }
 
+func (ndb *nodeDB) traverseStateChanges(startVersion, endVersion int64, fn func(version int64, changeSet *ChangeSet) error) error {
```
Comment (tac0turtle) [inline nodedb.go:1045]: missing godoc

File: iterator.go [hunk 3]
```diff
@@ -259,3 +259,74 @@ func (iter *Iterator) Error() error {
 func (iter *Iterator) IsFast() bool {
 	return false
 }
+
+// NodeIterator is an iterator for nodeDB to traverse a tree.
```
Comment (tac0turtle) [inline iterator.go:263]: can we add a bit more on the godoc, is the caller responsible for anything, etc..

File: diff.go [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+	Pairs []KVPair
+}
+
+type KVPair struct {
+	Delete bool
+	Key    []byte
+	Value  []byte
+}
+
+func StateChanges(ndb *nodeDB, version int64, root []byte, successorRoot []byte) (*ChangeSet, error) {
```
Comment (tac0turtle) [inline diff.go:21]: godocs please

**Comment by yihuang** [comment]
godoc added, can we backport to 0.19.x? it's an additive feature.

**Review by tac0turtle — APPROVED** [review]
utACK. I think its fine to backport

**Comment by cool-develope** [comment]
could you provide some covering tests?

**Comment by yihuang** [comment]
> could you provide some covering tests?

done, added randomized round-trip test case to make sure extracted change set is identical to the original ones.

**Review by cool-develope — APPROVED** [review]

**Comment by tac0turtle** [comment]
@mergifyio backport release/v0.19.0

**Comment by yihuang** [comment]
awesome 👍

**Comment by yihuang** [comment]
Sorry, I delete the branch too fast.

**Comment by tac0turtle** [comment]
@mergifyio backport release/v0.19.x

**Pull request merged**